### DISCUSSION TITLE
Accessing a TMap class

### DISCUSSION THREAD

**Discussion by Renal-Of-Loon** [scikit-hep/uproot5#1069] [Q&A] [answered]
Greetings all,

I am trying to access some data in a root file which is under the form of a TMap (named "parameters"). When trying to access it with `myfile["parameters"]` I only get `Unknown TMap at 0x.....`, using `myfile["parameters"].all_members` returns a 0 length dict.

The class does however appear in `myfile.classnames()` as `'parameters;1':'TMap'`

Is there a specific way to grab data from this structure or no with the library?

Thank you kindly.

**Comment by jpivarski** [comment]
"Unknown" means it couldn't find StreamerInfo for this type or otherwise couldn't get the information about how to deserialize it. (The type names, e.g. in `myfile.classnames()`, are always available.)

What does [TFile::ShowStreamerInfo](https://root.cern.ch/doc/master/classTFile.html#aa57bbc608f1e573cccb2c5637e441cfc) show you in ROOT? Is the `TMap` definition among the streamers of that file?

**Reply by Renal-Of-Loon** [reply]
It doesn't seem to be in there, I've made a copy-paste of the output [on pastebin](https://pastebin.com/hkd7kcAU)

Would this be something missing in the generation of the ROOT file? I believe the relevant lines in the generation of the root file should be the following:
```c++
#include "TMap.h"
//...
parameters_ = new TMap();
//...
void SimulationPlan::setParametersTMap(Controls *controls) {
  for (auto i : controls->vm) {
    TObjString *s1 = new TObjString(i.first.c_str());
    TObjString *s2 =
        new TObjString(controls->parameterToString(i.first).c_str());
    parameters_->Add((TObject *)s1, (TObject *)s2);
  }
}
//...
TFile *d = new TFile(outputFile_.c_str(), "recreate");
//...
d->WriteObject(parameters_, "parameters");
```
I realize this is very fragmented, but the goal was simply to create a Key:value map for some simulation parameters

**Reply by jpivarski** [reply]
I see that the `TMap` is not in the streamers, so the file is not fully self-describing. ROOT has `TMap` built in, so this missing streamer may be overlooked. I don't know how to coax ROOT to include that streamer in the file, either. I've seen similar issues like this, usually involving hadd.

I see that all of your key-value pairs are strings. You can put a `TObjString` directly into a directory. Suppose you write the `parameters` as JSON? Then you can use `json.loads` to decode it on the Python side.

Incidentally, `TObjString` is one of the few types that Uproot can also write. I used it as a test example in development because it's so simple.

**Reply by Renal-Of-Loon** [reply]
These values do end up eventually as a python dictionary when pulled from the root file yeah (ultimately stored in another format but it does go through json in the process). The origin of the parameters actually come in part from a BOOST parameter file, and others are computed in-process. The idea was to keep everything in just the one .root file to avoid many files hanging around.

I am not entirely certain what you mean by "You can put a TObjString directly into a directory" however. I imagine this is to do with a TDirectory? I'm afraid I'm not overly familiar with ROOT beyond the very (very) basics.

**Reply by jpivarski** [reply, marked as answer]
You used

```c++
d->WriteObject(parameters_, "parameters");
```

to write the `parameters_` TMap in directory `d` with name `"parameters"`. If `parameters_` were a `TObjString`, I think you would be able to write that just as easily.

I just checked, and yes you can:

```c++
root [0] TFile* d = new TFile("tmp.root", "RECREATE")
(TFile *) 0x55ebe0364bb0
root [1] TObjString* parameters_ = new TObjString("hello")
(TObjString *) 0x55ebe19ad570
root [2] d->WriteObject(parameters_, "parameters")
(int) 95
root [3] d->Close()
```

```python
>>> import uproot
>>> file = uproot.open("tmp.root")
>>> file["parameters"]
<TObjString 'hello' at 0x7fd911854c80>
```

So if instead of writing `"hello"`, you wrote a JSON document, you'd be able to load that as a Python dict with `json.loads`.

**Reply by Renal-Of-Loon** [reply]
Oh that's a great idea. Thank you kindly for the work and effort, I really appreciate it!

**Reply by Renal-Of-Loon** [reply]
I just figured I'd follow up for closure. This did end up working with minimal rework and entirely as expected. Thank you again for the easy workaround and clear example; greatly appreciate it!

**Comment by JacekHoleczek** [comment]
[No streamer info saved for TMaps?](https://root-forum.cern.ch/t/no-streamer-info-saved-for-tmaps/32728)

**Reply by jpivarski** [reply]
Thanks for the reference! Okay, then this is one that we could hard-code, like TList.

**Reply by JacekHoleczek** [reply]
BTW. You may also be interested in this: [`StreamerInfo is not stored in TFile with "update" mode`](https://root-forum.cern.ch/t/streamerinfo-is-not-stored-in-tfile-with-update-mode/57383)
And maybe also this: [`"Error in <TList::Clear>: A list is accessing an object already deleted (list name = TList)" when opening a file created by ROOT 6.30, using ROOT 6.14.09`](https://root-forum.cern.ch/t/error-in-tlist-clear-a-list-is-accessing-an-object-already-deleted-list-name-tlist-when-opening-a-file-created-by-root-6-30-using-root-6-14-09/57588)
